File: app/src/main/java/com/pij/zworkout/workout/viewmodel/CreateWorkoutFeature.java
package com.pij.zworkout.workout.viewmodel;

import android.support.annotation.NonNull;

import com.pij.horrocks.Interaction;
import com.pij.horrocks.Reducer;
import com.pij.zworkout.workout.State;

import javax.inject.Provider;

/**
 * <p>Created on 01/03/2018.</p>
 *
 * @author Pierrejean
 */

public class CreateWorkoutFeature implements Interaction<Object, State> {
    private final Provider<String> name;

    public CreateWorkoutFeature(Provider<String> name) {
        this.name = name;
    }

    @NonNull
    @Override
    public Reducer<State> process(@NonNull Object event) {
        return current -> current.toBuilder()
                .workout(current.workout().name(name.get()))
                .file(current.file().toBuilder().name(name.get()).build())
                .build();
    }

}

File: app/src/main/java/com/pij/zworkout/workout/viewmodel/CreateWorkoutFeature.kt
package com.pij.zworkout.workout.viewmodel

import com.annimon.stream.Optional
import com.pij.horrocks.Interaction
import com.pij.horrocks.Reducer
import com.pij.zworkout.workout.State
import java.io.File

import javax.inject.Provider

/**
 *
 * Created on 01/03/2018.
 *
 * @author Pierrejean
 */

class CreateWorkoutFeature(private val name: Provider<String>) : Interaction<Any, State> {

    override fun process(event: Any): Reducer<State> {
        return Reducer { current ->
            current.toBuilder()
                    .workout(current.workout().name(name.get()))
                    .file(Optional.empty<File>())
                    .nameIsReadOnly(false)
                    .build()
        }
    }

}
